### PR TITLE
Updated Base Pointer Controller Pointer Name Property  …

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
- Added a backing field, in case the gameObject reference is destroyed when attempting to get the pointer name

### CODE REVIEW

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 1]
```diff
@@ -190,11 +190,17 @@ public uint PointerId
...
         /// <inheritdoc />
         public string PointerName
         {
-            get { return gameObject.name; }
-            set { gameObject.name = value; }
+            get { return pointerName; }
+            set
+            {
+                pointerName = value;
+                gameObject.name = value;
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: This will still null ref if the gameObject is destroyed.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: Yes, but we only set the name once we create the pointer, so I doubt that can ever really happen.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]:
```suggestion
                if(gameObject == null)
                {
                    Debug.LogError($"{pointerName}'s GameObject reference is missing!");
                    return;
                }
                
                gameObject.name = value;
```

**Comment by StephenHodgson** [comment]
>isn't this hiding an issue? if the underlying gameobject has been destroyed, the pointer is no longer valid... how do you envision an app getting into this situation?

I've never had an issue _setting_ the name, only _getting_ it during shutdown because of the non-deterministic nature of unity's game loop. For example, if a GameObject get's destroyed before the MixedRealityToolkit ServiceLocator finished shutting down the InputSystem.

**Comment by keveleigh** [comment]
>For example, if a GameObject get's destroyed before the MixedRealityToolkit ServiceLocator finished shutting down the InputSystem.

If the `gameObject` is destroyed, won't the pointer script also be destroyed? Does checking the pointer script reference for null before trying to access the name not work?

**Comment by StephenHodgson** [comment]
>If the gameObject is destroyed, won't the pointer script also be destroyed? Does checking the pointer script reference for null before trying to access the name not work?

No it doesn't which is why we're getting the issue in the first place. 😅

**Comment by keveleigh** [comment]
Which line is causing the issue?

**Comment by StephenHodgson** [comment]
The `gameObject` can sometimes be null [when exiting play mode in the editor](
https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2944/files#diff-820aa912c19ab5fb36b99791f204016cL196) because it is cleaned up before the system shuts down.

**Comment by keveleigh** [comment]
Right, that part is clear from this PR. I was asking if there's a line in the Toolkit asking the pointer for its name at a time when the `gameObject` is null.

**Comment by StephenHodgson** [comment]
When we're [unregistering the pointers and removing it from the list](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit-SDK/Features/Input/FocusProvider.cs#L494). Under the [hood the `Remove` function tries to get the hash cod](https://referencesource.microsoft.com/#system.core/system/Collections/Generic/HashSet.cs,289)e.  In our [hash code override](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs#L368) we access [the property](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs#L196), which returns the `gameObject.name` instead of a backed field.

**Comment by davidkline-ms** [comment]
@StephenHodgson, that's a great description of the issue. Thanks for adding it. Makes the change much easier to understand.

**Pull request merged**